src/xenia/kernel/xam/xam_user.cc

```diff
@@ -552,23 +552,70 @@ dword_result_t XamShowSigninUI(dword_t unk, dword_t unk_mask) {
 }
 DECLARE_XAM_EXPORT1(XamShowSigninUI, kUserProfiles, kStub);
 
+#pragma pack(push, 1)
+struct X_ACHIEVEMENT_DETAILS {
+  xe::be<uint32_t> id;
+  xe::be<uint32_t> label_ptr;
+  xe::be<uint32_t> description_ptr;
+  xe::be<uint32_t> unachieved_ptr;
+  xe::be<uint32_t> image_id;
+  xe::be<uint32_t> gamerscore;
+  xe::be<uint64_t> unlock_time;
+  xe::be<uint32_t> flags;
+};
+static_assert_size(X_ACHIEVEMENT_DETAILS, 36);
+#pragma pack(pop)
+
 dword_result_t XamUserCreateAchievementEnumerator(dword_t title_id,
                                                   dword_t user_index,
                                                   dword_t xuid, dword_t flags,
                                                   dword_t offset, dword_t count,
                                                   lpdword_t buffer_size_ptr,
                                                   lpdword_t handle_ptr) {
+  if (!count || !buffer_size_ptr || !handle_ptr) {
+    return X_ERROR_INVALID_PARAMETER;
+  }
+
+  if (user_index >= 4) {
+    return X_ERROR_INVALID_PARAMETER;
+  }
+
+  // Providing flag 0x20 only gives you basic info about achievements
+  uint32_t entry_size = 500;
+  if (flags & 0x20) {
+    entry_size = sizeof(X_ACHIEVEMENT_DETAILS);
+  }
+
   if (buffer_size_ptr) {
-    *buffer_size_ptr = 500 * count;
+    *buffer_size_ptr = entry_size * count;
   }
 
   auto e = object_ref<XStaticEnumerator>(
-      new XStaticEnumerator(kernel_state(), count, 500));
+      new XStaticEnumerator(kernel_state(), count, entry_size));
   auto result = e->Initialize(user_index, 0xFB, 0xB000A, 0xB000B, 0);
   if (XFAILED(result)) {
     return result;
   }
 
+  std::u16string dummy_achievement_text = to_utf16("Dummy Text");
+  uint32_t alloc_ptr =
+      kernel_memory()->SystemHeapAlloc(sizeof(dummy_achievement_text));
```

Comment:
You're allocating heap memory here, any dummy text should be within the the buffer provided to the function, and guest pointers should point to that.

Reply:
I kindly disagree.
It depends on flags provided to this function.

If you only provides flag 0x20 then you should get pointer from SPA to certain label. (that is why I had to add this allocation)

In other cases (iirc flags 0x2, 0x4, 0x8) yes, there we have space in provided buffer to store these strings and they're stored right after all X_ACHIEVEMENTS_DETAILS structures.

Reply:
What's "SPA"? If the flags control strings being present, why would you not have space for them in the user-provided buffer?

Reply:
If it really needs empty space outside of the user-provided buffer for some reason (why wouldn't the pointers just be NULL if there's nothing?), then the allocation should be moved to when the module is created, not for every call to creating the enumerator.

Reply:
SPA contains information about achievements, stats etc. and this is loaded into memory while game starts. 
Emoose knows more about how everything works around it as they wrote code for it long time ago.

I can agree that we can remove it from it just in case, when game abuses this function. 
The pointer cannot be null as games assume that this is always loaded correctly (for example Red Dead Redemption does that)

Reply:
Ah ok. You previously said RDR passes every flag possible (-1). Shouldn't that mean it is also providing buffer space for strings? Or is it not?

Reply:
Yeah it does. I haven't checked how it is handled, but I guess it contains space to store these strings after achievements structure. 
However header still expect that we will pass pointer to location where it is stored instead of null.

Game doesn't have direct access to these strings. It takes pointer from this header.

Reply:
Yeah that's what I mean, the expectation would be that the strings would get copied after the achievement structs, and the structures would be filled out to pointers to these locations in the buffer as normal?

Reply:
Yes exactly, but only if you provide these specific flags (0x2, 0x4, 0x8 or 0x4, 0x8, 0x10)